src/libstd/sys_common/backtrace.rs

```diff
@@ -66,7 +66,14 @@ unsafe fn _print(w: &mut dyn Write, format: PrintFmt) -> io::Result<()> {
 }
 
 unsafe fn _print_fmt(fmt: &mut fmt::Formatter<'_>, print_fmt: PrintFmt) -> fmt::Result {
-    let cwd = env::current_dir().ok();
+    // Always 'fail' to get the cwd when running under Miri -
+    // this allows Miri to display backtraces in isolation mode
```

Comment:
This is a hack though, it'll apply even when not in isolation mode...
IMO it would make more sense to return a permission failure to the interpreted program in isolation mode. This is what also happens in "real" sandboxes.

Reply:
I'm concerned that this would make isolation mode almost useless. Any code which panics *anywhere* will be impossible to run in isolation mode. This means, for example, that you cannot run a program's test-suite in isolation mode if there are any `#[should_panic]` tests, which strikes me as an enormous limitation.

Reply:
In the future, we could add a more fine-grained isolation mode to Miri (e.g. allow the user to enable specific foreign functions). Until we have that, I don't think slightly better backtraces are worth imposing severe limitations on the programs that can be run in isolation mode.

Reply:
> Any code which panics anywhere will be impossible to run in isolation mode.

No, why that? I am suggesting to make `getcwd` return `-1` such that `env::current_dir().ok()` ends up as `None` without a code change here.

Reply:
Oh, I see.

I think this came up before - I'm not a huge fan of returning error codes in isolation mode - at least, not without explicit opt-in from the user.

This feels similar to `getrandom` - we should either give the user an explicit error (with a callstack) when an unsupported function is called, or have a command-line flag to opt in to using the external environment. I feel that returing an error code could make things incredibly confusing for users of Miri - the behavhior of the program could subtly change, without any indication that `getcwd` is to blame.

Reply:
> I think this came up before - I'm not a huge fan of returning error codes in isolation mode - at least, not without explicit opt-in from the user.

I am not the biggest fan either, but it strikes me as better as this alternative of making `_print_fmt` depend on Miri. As I mentioned, this is also standard for sandboxes.

> I feel that returning an error code could make things incredibly confusing for users of Miri - the behavior of the program could subtly change, without any indication that getcwd is to blame.

The same is true with `cfg(miri)` though, it introduces IMO much worse subtle differences between normal and Miri execution.

And returning error code is generally what sandboxes do, so I don't foresee much confusion. Either the program will panic/propagate the error, or it'll properly handle it. Both seem fine to me.

@oli-obk any thoughts?

Reply:
@RalfJung: My main concern here is consistency. I think that `getrandom` and `getcwd` should do the same thing in isolation - either abort the program with a backtrace, or generate an error code.

Reply:
That is a good point, and I agree we should then use error codes for the other functions as well. We can emit warnings when that happens. But we don't have an infrastructure for warnings yet...

I'd be fine with a temporary inconsistency, but if you prefer a temporary degrading in backtraces for non-isolated mode, that seems fine, too. I don't even know why this function needs to know the working directory.^^

So what about this: for now, we go with your approach, but please open an issue in Miri summarizing our discussion here and tracking that we want to eventually move to returning error codes from operations that fail due to isolation (unless @oli-obk thinks that's bad).

Reply:
Sounds good. I opened https://github.com/rust-lang/miri/issues/1034